### PR TITLE
ENH Adding categorical parallel_coordinates implementation

### LINKED ISSUE #12341: Categorical data not supported in plots of parallel coordinates
Just hoping to see if it might be possible to support categorical data in the fantastic parallel coordinates plot function.

I've uploaded an example dataset where the current code fails:

[Archive.zip](https://github.com/pydata/pandas/files/131996/Archive.zip)

Also, variable line widths that reflect frequency would also be highly appreciated. :)

### PR DESCRIPTION
**Pull request by stangirala**
- [x] closes #12341
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ ] whatsnew entry

Initial PR for #12341 

I have a couple of questions, the signature is different from the parallel_coordinates method and I was not sure if we need all of them (for example axvlines)?

I've added some tests but need suggestions on how to test _connect_polyline(). I was thinking of simply using a mock object for the axes objects and verifying their attributes were being accessed.

Some context, I have a new method called parallel_sets() which converts all of its inputs to a string expecting categorical data always, via a string_to_integer dictionary, and plots those integers.

### CODE REVIEW

**Comment by stangirala** [comment]
Not sure why https://travis-ci.org/pandas-dev/pandas/jobs/228503476 is failing on TestClipboard.test_round_trip_frame_sep, I'm not able to see the failure on my local environment.

**Comment by TomAugspurger** [comment]
I restarted it; that test does occasionally fail. Haven't been able to debug it yet.

File: pandas/tests/plotting/test_misc.py [hunk 1]
```diff
@@ -263,6 +263,47 @@ def test_parallel_coordinates_with_sorted_labels(self):
             # lables and colors are ordered strictly increasing
             assert prev[1] < nxt[1] and prev[0] < nxt[0]
 
+    def test_parallel_sets(self):
+        from functools import reduce
```
Comment (jreback) [inline pandas/tests/plotting/test_misc.py]: import at the top of the file.

File: pandas/tests/plotting/test_misc.py [hunk 2]
```diff
@@ -263,6 +263,47 @@ def test_parallel_coordinates_with_sorted_labels(self):
             # lables and colors are ordered strictly increasing
             assert prev[1] < nxt[1] and prev[0] < nxt[0]
 
+    def test_parallel_sets(self):
```
Comment (jreback) [inline pandas/tests/plotting/test_misc.py]: more as ``@slow``

File: pandas/plotting/_misc.py [hunk 3]
```diff
@@ -500,6 +500,175 @@ def parallel_coordinates(frame, class_column, cols=None, ax=None, color=None,
...
+    >>> parallel_sets(df, 'Name', color=('#556270',
+                             '#4ECDC4', '#C7F464'))
+    >>> plt.show()
+    """
+    parsets = _ParSets(color, colormap)
+    return parsets.par_sets(frame, class_column)
+
+
+class _ParSets(object):
+    def __init__(self, color=None, colormap=None):
```
Comment (jreback) [inline pandas/plotting/_misc.py]: can you fill out the doc-string

File: pandas/plotting/_misc.py [hunk 4]
```diff
@@ -500,6 +500,175 @@ def parallel_coordinates(frame, class_column, cols=None, ax=None, color=None,
     return ax
 
 
+def parallel_sets(frame, class_column, color=None, colormap=None, **kwds):
+    """Parallel sets plotting.
```
Comment (jreback) [inline pandas/plotting/_misc.py]: can you add a 1-liner (or more) about the utility of this

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
This looks fine. @TomAugspurger , @jorisvandenbossche 

side issue is we should create an issue to move all of the interesting plots to an external lib, maybe ``pandas-plotting`` (I mean the ones in ``pandas.plotting``)

**Comment by stangirala** [comment]
@jreback Added a few more tests, and one change, I thought the class labels should be in the legend but think it's a better idea to just leave it as an axes in the plot similar to http://kosara.net/papers/2006/Kosara_TVCG_2006.pdf

File: pandas/plotting/_misc.py (hunk 4 above)
Comment (stangirala) [inline pandas/plotting/_misc.py]: Added

**Comment by shoyer** [comment]
@stangirala can you please show an example of the visualizations this function produces and the code required?

**Comment by stangirala** [comment]
@shoyer Try this,
~~~~
import pandas as pd
import matplotlib.pyplot as plt
df = pd.DataFrame({'1':['abc', 'b', 'c', 'b'], '2':['d', 'f', 'g', 'd'], \
        '3':['h', 'e', 'j', 'h'], 'four': ['x', 'y', 'x', 'y']})
from pandas.plotting import parallel_sets
parallel_sets(df, "3")
plt.show()
~~~~

You should get this plot,
screen shot 2017-05-07 at 22 39 35
,

**Comment by TomAugspurger** [comment]
Sorry for lettings this sit for so long @stangirala 

Quick question: Can you explain the difference between parallel_coordinates and parallel_sets? Or let me know if this explanation is correct:

- Parallel coordinates plots one line per observation across various dimensions of the data
- Parallel sets plots an *aggregation* of the data, across various dimensions, where the width of each polygon is proportional to the frequency of that value in the dataset?

**Comment by stangirala** [comment]
@TomAugspurger (Was on vacation :) Yes, your explanation is correct. The import distinction is the notion of frequency being captured with parallel sets. I think I made a mistake of using the term parallel sets to distinguish from the parallel coordinates function given the ability to handle categorical data (a fixed size set as opposed to a number set). If I fix the title and the method name it should be clearer. Sorry about the confusion.

**Comment by stangirala** [comment]
@TomAugspurger If that makes sense, I was thinking of replacing the existing parallel coordinates function with the new one.

**Comment by TomAugspurger** [comment]
So the advantage of your implementation over the current one is that it handles non-numeric data, correct?

> @TomAugspurger If that makes sense, I was thinking of replacing the existing parallel coordinates function with the new one.

You can give it a shot and see which tests fail. I suspect that they all will. If I had to guess, the output (visual, and actual objects returned) are too different to simply replace it. We'd probably want to deprecate the old way and replace it with yours, assuming there are no downsides (do you see any?)

**Comment by jorisvandenbossche** [comment]
>  We'd probably want to deprecate the old way and replace it with yours, assuming there are no downsides (do you see any?)

I think the current behaviour when having only numerical columns should (more or less) just stay the same? But adding the ability to also use the function with non-numerical data. 
One question is how to do the y-axis (as this one is now shared for numerical columns). But also for numerical columns, having the option to *not* share the y-axis can be useful I think.

**Comment by stangirala** [comment]
@TomAugspurger @jorisvandenbossche One way to keep the current implementation is we can have as many subplots as there are columns and squish them next to each other (https://matplotlib.org/faq/howto_faq.html#move-the-edge-of-an-axes-to-make-room-for-tick-labels)?

**Comment by jreback** [comment]
if you can update according to comments

**Comment by jreback** [comment]
closing as stale, if you'd like to continue, pls ping.

**Pull request closed without merging**